File: db4o.net/drs/Db4objects.Drs.Tests/Db4objects.Drs.Tests/ReplicationProviderTest.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System.Collections;
using Db4oUnit;
using Db4objects.Db4o;
using Db4objects.Db4o.Ext;
using Db4objects.Drs.Foundation;
using Db4objects.Drs.Inside;
using Db4objects.Drs.Tests;
using Db4objects.Drs.Tests.Data;

namespace Db4objects.Drs.Tests
{
	public class ReplicationProviderTest : DrsTestCase
	{
		protected byte[] BSignatureBytes;

		protected IReadonlyReplicationProviderSignature BSignature;

		private IReadonlyReplicationProviderSignature ASignature;

		public virtual void TestReplicationLifeCycle()
		{
			BSignatureBytes = B().Provider().GetSignature().GetSignature();
			ASignature = A().Provider().GetSignature();
			BSignature = B().Provider().GetSignature();
			TstObjectUpdate();
			TstSignature();
			TstObjectsChangedSinceLastReplication();
			TstReferences();
			TstStore();
			TstRollback();
			TstDeletion();
		}

		protected virtual void TstDeletion()
		{
			A().Provider().StoreNew(new Pilot("Pilot1", 42));
			Pilot o = new Pilot("Pilot2", 43);
			A().Provider().StoreNew(o);
			A().Provider().Commit();
			A().Provider().StoreNew(new Pilot("Pilot3", 44));
			A().Provider().Delete(FindPilot("Pilot1"));
			Car car = new Car("Car1");
			car.SetPilot(FindPilot("Pilot2"));
			A().Provider().StoreNew(car);
			A().Provider().Commit();
			StartReplication();
			IDrsUUID uuidCar1 = Uuid(FindCar("Car1"));
			Assert.IsNotNull(uuidCar1);
			A().Provider().ReplicateDeletion(uuidCar1);
			CommitReplication();
			Assert.IsNull(FindCar("Car1"));
			StartReplication();
			IDrsUUID uuidPilot2 = Uuid(FindPilot("Pilot2"));
			Assert.IsNotNull(uuidPilot2);
			A().Provider().ReplicateDeletion(uuidPilot2);
			CommitReplication();
			Assert.IsNull(FindPilot("Pilot2"));
		}

		private void CommitReplication()
		{
			A().Provider().CommitReplicationTransaction();
			B().Provider().CommitReplicationTransaction();
		}

		private object FindCar(string model)
		{
			IEnumerator cars = A().Provider().GetStoredObjects(typeof(Car)).GetEnumerator();
			while (cars.MoveNext())
			{
				Car candidate = (Car)cars.Current;
				if (candidate.GetModel().Equals(model))
				{
					return candidate;
				}
			}
			return null;
		}

		private Pilot FindPilot(string name)
		{
			IEnumerator pilots = A().Provider().GetStoredObjects(typeof(Pilot)).GetEnumerator
				();
			while (pilots.MoveNext())
			{
				Pilot candidate = (Pilot)pilots.Current;
				if (candidate.Name().Equals(name))
				{
					return candidate;
				}
			}
			return null;
		}

		private SPCChild GetOneChildFromA()
		{
			return GetOneChild(A());
		}

		private SPCChild GetOneChild(IDrsProviderFixture fixture)
		{
			IObjectSet storedObjects = fixture.Provider().GetStoredObjects(typeof(SPCChild));
			Assert.AreEqual(1, storedObjects.Count);
			IEnumerator iterator = storedObjects.GetEnumerator();
			Assert.IsTrue(iterator.MoveNext());
			return (SPCChild)iterator.Current;
		}

		private void StartReplication()
		{
			A().Provider().StartReplicationTransaction(BSignature);
			B().Provider().StartReplicationTransaction(ASignature);
		}

		private void TstObjectUpdate()
		{
			SPCChild child = new SPCChild("c1");
			A().Provider().StoreNew(child);
			A().Provider().Commit();
			StartReplication();
			SPCChild reloaded = GetOneChildFromA();
			long oldVer = A().Provider().ProduceReference(reloaded, null, null).Version();
			CommitReplication();
			SPCChild reloaded2 = GetOneChildFromA();
			reloaded2.SetName("c3");
			//System.out.println("==============BEGIN DEBUG");
			A().Provider().Update(reloaded2);
			A().Provider().Commit();
			//System.out.println("==============END DEBUG");
			StartReplication();
			SPCChild reloaded3 = GetOneChildFromA();
			long newVer = A().Provider().ProduceReference(reloaded3, null, null).Version();
			CommitReplication();
			Assert.IsGreater(oldVer, newVer);
		}

		private void TstObjectsChangedSinceLastReplication()
		{
			Pilot object1 = new Pilot("John Cleese", 42);
			Pilot object2 = new Pilot("Terry Gilliam", 53);
			Car object3 = new Car("Volvo");
			A().Provider().StoreNew(object1);
			A().Provider().StoreNew(object2);
			A().Provider().StoreNew(object3);
			A().Provider().Commit();
			StartReplication();
			IObjectSet changed = A().Provider().ObjectsChangedSinceLastReplication();
			Assert.AreEqual(3, changed.Count);
			IObjectSet os = A().Provider().ObjectsChangedSinceLastReplication(typeof(Pilot));
			Assert.AreEqual(2, os.Count);
			IEnumerator pilots = os.GetEnumerator();
			//		Assert.isTrue(pilots.contains(findPilot("John Cleese")));
			//	Assert.isTrue(pilots.contains(findPilot("Terry Gilliam")));
			IEnumerator cars = A().Provider().ObjectsChangedSinceLastReplication(typeof(Car))
				.GetEnumerator();
			Assert.AreEqual("Volvo", ((Car)Next(cars)).GetModel());
			Assert.IsFalse(cars.MoveNext());
			CommitReplication();
			StartReplication();
			Assert.IsFalse(A().Provider().ObjectsChangedSinceLastReplication().GetEnumerator(
				).MoveNext());
			CommitReplication();
			Pilot pilot = (Pilot)Next(A().Provider().GetStoredObjects(typeof(Pilot)).GetEnumerator
				());
			pilot.SetName("Terry Jones");
			Car car = (Car)Next(A().Provider().GetStoredObjects(typeof(Car)).GetEnumerator());
			car.SetModel("McLaren");
			A().Provider().Update(pilot);
			A().Provider().Update(car);
			A().Provider().Commit();
			StartReplication();
			Assert.AreEqual(2, A().Provider().ObjectsChangedSinceLastReplication().Count);
			pilots = A().Provider().ObjectsChangedSinceLastReplication(typeof(Pilot)).GetEnumerator
				();
			Assert.AreEqual("Terry Jones", ((Pilot)Next(pilots)).Name());
			Assert.IsFalse(pilots.MoveNext());
			cars = A().Provider().ObjectsChangedSinceLastReplication(typeof(Car)).GetEnumerator
				();
			Assert.AreEqual("McLaren", ((Car)Next(cars)).GetModel());
			Assert.IsFalse(cars.MoveNext());
			CommitReplication();
			A().Provider().DeleteAllInstances(typeof(Pilot));
			A().Provider().DeleteAllInstances(typeof(Car));
			A().Provider().Commit();
		}

		private object Next(IEnumerator iterator)
		{
			Assert.IsTrue(iterator.MoveNext());
			return iterator.Current;
		}

		private void TstReferences()
		{
			Pilot pilot = new Pilot("tst References", 42);
			A().Provider().StoreNew(pilot);
			A().Provider().Commit();
			StartReplication();
			Pilot object1 = (Pilot)Next(A().Provider().GetStoredObjects(typeof(Pilot)).GetEnumerator
				());
			IReplicationReference reference = A().Provider().ProduceReference(object1, null, 
				null);
			Assert.AreEqual(object1, reference.Object());
			IDrsUUID uuid = reference.Uuid();
			IReplicationReference ref2 = A().Provider().ProduceReferenceByUUID(uuid, typeof(Pilot
				));
			Assert.AreEqual(reference, ref2);
			A().Provider().ClearAllReferences();
			IDrsUUID db4oUUID = A().Provider().ProduceReference(object1, null, null).Uuid();
			Assert.AreEqual(uuid, db4oUUID);
			CommitReplication();
			A().Provider().DeleteAllInstances(typeof(Pilot));
			A().Provider().Commit();
		}

		private void TstRollback()
		{
			if (!A().Provider().SupportsRollback())
			{
				return;
			}
			if (!B().Provider().SupportsRollback())
			{
				return;
			}
			StartReplication();
			Pilot object1 = new Pilot("Albert Kwan", 25);
			IDrsUUID uuid = new DrsUUIDImpl(new Db4oUUID(5678, BSignatureBytes));
			IReplicationReference @ref = new ReplicationReferenceImpl(object1, uuid, 1);
			A().Provider().ReferenceNewObject(object1, @ref, null, null);
			A().Provider().StoreReplica(object1);
			Assert.IsFalse(A().Provider().WasModifiedSinceLastReplication(@ref));
			A().Provider().RollbackReplication();
			A().Provider().StartReplicationTransaction(BSignature);
			Assert.IsNull(A().Provider().ProduceReference(object1, null, null));
			IReplicationReference byUUID = A().Provider().ProduceReferenceByUUID(uuid, object1
				.GetType());
			Assert.IsNull(byUUID);
			A().Provider().RollbackReplication();
			B().Provider().RollbackReplication();
		}

		private void TstSignature()
		{
			Assert.IsNotNull(A().Provider().GetSignature());
		}

		private void TstStore()
		{
			StartReplication();
			Pilot object1 = new Pilot("John Cleese", 42);
			IDrsUUID uuid = new DrsUUIDImpl(new Db4oUUID(15, BSignatureBytes));
			IReplicationReference @ref = new ReplicationReferenceImpl("ignoredSinceInOtherProvider"
				, uuid, 1);
			A().Provider().ReferenceNewObject(object1, @ref, null, null);
			A().Provider().StoreReplica(object1);
			IReplicationReference reference = A().Provider().ProduceReferenceByUUID(uuid, object1
				.GetType());
			Assert.AreEqual(reference, A().Provider().ProduceReference(object1, null, null));
			Assert.AreEqual(object1, reference.Object());
			CommitReplication();
			StartReplication();
			IEnumerator storedObjects = A().Provider().GetStoredObjects(typeof(Pilot)).GetEnumerator
				();
			Pilot reloaded = (Pilot)Next(storedObjects);
			Assert.IsFalse(storedObjects.MoveNext());
			reference = A().Provider().ProduceReferenceByUUID(uuid, object1.GetType());
			Assert.AreEqual(reference, A().Provider().ProduceReference(reloaded, null, null));
			reloaded.SetName("i am updated");
			A().Provider().StoreReplica(reloaded);
			A().Provider().ClearAllReferences();
			CommitReplication();
			StartReplication();
			reference = A().Provider().ProduceReferenceByUUID(uuid, reloaded.GetType());
			Assert.AreEqual("i am updated", ((Pilot)reference.Object()).Name());
			CommitReplication();
			A().Provider().DeleteAllInstances(typeof(Pilot));
			A().Provider().Commit();
		}

		private IDrsUUID Uuid(object obj)
		{
			return A().Provider().ProduceReference(obj, null, null).Uuid();
		}
	}
}

File: drs/src/test/com/db4o/drs/test/ReplicationProviderTest.java
/* Copyright (C) 2004 - 2008  Versant Inc.  http://www.db4o.com

This file is part of the db4o open source object database.

db4o is free software; you can redistribute it and/or modify it under
the terms of version 2 of the GNU General Public License as published
by the Free Software Foundation and as clarified by db4objects' GPL 
interpretation policy, available at
http://www.db4o.com/about/company/legalpolicies/gplinterpretation/
Alternatively you can write to db4objects, Inc., 1900 S Norfolk Street,
Suite 350, San Mateo, CA 94403, USA.

db4o is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA. */
package com.db4o.drs.test;

import java.util.*;

import com.db4o.*;
import com.db4o.drs.foundation.*;
import com.db4o.drs.inside.*;
import com.db4o.drs.test.data.*;
import com.db4o.ext.*;

import db4ounit.*;


public class ReplicationProviderTest extends DrsTestCase {

	protected byte[] B_SIGNATURE_BYTES;
	protected ReadonlyReplicationProviderSignature B_SIGNATURE;
	private ReadonlyReplicationProviderSignature A_SIGNATURE;

	public void testReplicationLifeCycle() {
		B_SIGNATURE_BYTES = b().provider().getSignature().getSignature();

		A_SIGNATURE = a().provider().getSignature();
		B_SIGNATURE = b().provider().getSignature();

		tstObjectUpdate();

		tstSignature();

		tstObjectsChangedSinceLastReplication();

		tstReferences();

		tstStore();

		tstRollback();

		tstDeletion();
	}


	protected void tstDeletion() {
		a().provider().storeNew(new Pilot("Pilot1", 42));
		Pilot o = new Pilot("Pilot2", 43);
		a().provider().storeNew(o);

		a().provider().commit();

		a().provider().storeNew(new Pilot("Pilot3", 44));

		a().provider().delete(findPilot("Pilot1"));

		Car car = new Car("Car1");
		car.setPilot(findPilot("Pilot2"));
		a().provider().storeNew(car);

		a().provider().commit();

		startReplication();

		DrsUUID uuidCar1 = uuid(findCar("Car1"));
		Assert.isNotNull(uuidCar1);

		a().provider().replicateDeletion(uuidCar1);

		commitReplication();

		Assert.isNull(findCar("Car1"));

		startReplication();

		DrsUUID uuidPilot2 = uuid(findPilot("Pilot2"));
		Assert.isNotNull(uuidPilot2);
		a().provider().replicateDeletion(uuidPilot2);

		commitReplication();

		Assert.isNull(findPilot("Pilot2"));

	}

	private void commitReplication() {
		a().provider().commitReplicationTransaction();
		b().provider().commitReplicationTransaction();
	}

	private Object findCar(String model) {
		Iterator cars = a().provider().getStoredObjects(Car.class).iterator();
		while (cars.hasNext()) {
			Car candidate = (Car) cars.next();
			if (candidate.getModel().equals(model)) return candidate;
		}
		return null;
	}

	private Pilot findPilot(String name) {
		Iterator pilots = a().provider().getStoredObjects(Pilot.class).iterator();
		while (pilots.hasNext()) {
			Pilot candidate = (Pilot) pilots.next();
			if (candidate.name().equals(name)) return candidate;
		}
		return null;
	}

	private SPCChild getOneChildFromA() {
		return getOneChild(a());
	}

	private SPCChild getOneChild(DrsProviderFixture fixture) {
		ObjectSet storedObjects = fixture.provider().getStoredObjects(SPCChild.class);
		Assert.areEqual(1, storedObjects.size());
		
		Iterator iterator = storedObjects.iterator();
		Assert.isTrue(iterator.hasNext());
		return (SPCChild) iterator.next();
	}

	private void startReplication() {
		a().provider().startReplicationTransaction(B_SIGNATURE);
		b().provider().startReplicationTransaction(A_SIGNATURE);
	}

	private void tstObjectUpdate() {
		SPCChild child = new SPCChild("c1");
		a().provider().storeNew(child);
		a().provider().commit();

		startReplication();
		SPCChild reloaded = getOneChildFromA();
		long oldVer = a().provider().produceReference(reloaded, null, null).version();
		commitReplication();

		SPCChild reloaded2 = getOneChildFromA();
		reloaded2.setName("c3");

		//System.out.println("==============BEGIN DEBUG");
		a().provider().update(reloaded2);
		a().provider().commit();
		//System.out.println("==============END DEBUG");

		startReplication();
		SPCChild reloaded3 = getOneChildFromA();
		long newVer = a().provider().produceReference(reloaded3, null, null).version();
		commitReplication();

		Assert.isGreater(oldVer, newVer);
	}

	private void tstObjectsChangedSinceLastReplication() {
		Pilot object1 = new Pilot("John Cleese", 42);
		Pilot object2 = new Pilot("Terry Gilliam", 53);
		Car object3 = new Car("Volvo");

		a().provider().storeNew(object1);
		a().provider().storeNew(object2);
		a().provider().storeNew(object3);

		a().provider().commit();

		startReplication();

		ObjectSet changed = a().provider().objectsChangedSinceLastReplication();
		Assert.areEqual(3, changed.size());

		ObjectSet os = a().provider().objectsChangedSinceLastReplication(Pilot.class);
		Assert.areEqual(2, os.size());
		
		Iterator pilots = os.iterator();
//		Assert.isTrue(pilots.contains(findPilot("John Cleese")));
	//	Assert.isTrue(pilots.contains(findPilot("Terry Gilliam")));
		
		Iterator cars = a().provider().objectsChangedSinceLastReplication(Car.class).iterator();		
		Assert.areEqual("Volvo", ((Car) next(cars)).getModel());
		Assert.isFalse(cars.hasNext());

		commitReplication();

		startReplication();

		Assert.isFalse(a().provider().objectsChangedSinceLastReplication().iterator().hasNext());
		commitReplication();

		Pilot pilot = (Pilot) next(a().provider().getStoredObjects(Pilot.class).iterator());
		pilot.setName("Terry Jones");

		Car car = (Car) next(a().provider().getStoredObjects(Car.class).iterator());
		car.setModel("McLaren");

		a().provider().update(pilot);
		a().provider().update(car);

		a().provider().commit();

		startReplication();

		Assert.areEqual(2, a().provider().objectsChangedSinceLastReplication().size());

		pilots = a().provider().objectsChangedSinceLastReplication(Pilot.class).iterator();
		Assert.areEqual("Terry Jones", ((Pilot) next(pilots)).name());
		Assert.isFalse(pilots.hasNext());

		cars = a().provider().objectsChangedSinceLastReplication(Car.class).iterator();		
		Assert.areEqual("McLaren", ((Car) next(cars)).getModel());
		Assert.isFalse(cars.hasNext());
		commitReplication();

		a().provider().deleteAllInstances(Pilot.class);
		a().provider().deleteAllInstances(Car.class);
		a().provider().commit();
	}


	private Object next(Iterator iterator) {
		Assert.isTrue(iterator.hasNext());
		return iterator.next();
	}

	private void tstReferences() {
		Pilot pilot = new Pilot("tst References", 42);
		a().provider().storeNew(pilot);
		a().provider().commit();

		startReplication();

		Pilot object1 = (Pilot) next(a().provider().getStoredObjects(Pilot.class).iterator());

		ReplicationReference reference = a().provider().produceReference(object1, null, null);
		Assert.areEqual(object1, reference.object());

		DrsUUID uuid = reference.uuid();
		ReplicationReference ref2 = a().provider().produceReferenceByUUID(uuid, Pilot.class);
		Assert.areEqual(reference, ref2);

		a().provider().clearAllReferences();
		DrsUUID db4oUUID = a().provider().produceReference(object1, null, null).uuid();
		Assert.areEqual(uuid, db4oUUID);
		commitReplication();

		a().provider().deleteAllInstances(Pilot.class);
		a().provider().commit();
	}

	private void tstRollback() {
		if (!a().provider().supportsRollback()) return;
		if (!b().provider().supportsRollback()) return;

		startReplication();

		Pilot object1 = new Pilot("Albert Kwan", 25);
		DrsUUID uuid = new DrsUUIDImpl(new Db4oUUID(5678, B_SIGNATURE_BYTES));

		ReplicationReference ref = new ReplicationReferenceImpl(object1, uuid, 1);
		a().provider().referenceNewObject(object1, ref, null, null);

		a().provider().storeReplica(object1);
		Assert.isFalse(a().provider().wasModifiedSinceLastReplication(ref));
		
		a().provider().rollbackReplication();

		a().provider().startReplicationTransaction(B_SIGNATURE);
		Assert.isNull(a().provider().produceReference(object1, null, null));
		ReplicationReference byUUID = a().provider().produceReferenceByUUID(uuid, object1.getClass());
		Assert.isNull(byUUID);
		
		a().provider().rollbackReplication();
		b().provider().rollbackReplication();
	}

	private void tstSignature() {
		Assert.isNotNull(a().provider().getSignature());
	}

	private void tstStore() {
		startReplication();

		Pilot object1 = new Pilot("John Cleese", 42);
		DrsUUID uuid = new DrsUUIDImpl(new Db4oUUID(15, B_SIGNATURE_BYTES));

		ReplicationReference ref = new ReplicationReferenceImpl("ignoredSinceInOtherProvider", uuid, 1);

		a().provider().referenceNewObject(object1, ref, null, null);
		
		a().provider().storeReplica(object1);
		ReplicationReference reference = a().provider().produceReferenceByUUID(uuid, object1.getClass());
		Assert.areEqual(reference, a().provider().produceReference(object1, null, null));
		Assert.areEqual(object1, reference.object());

		commitReplication();
		startReplication();

		Iterator storedObjects = a().provider().getStoredObjects(Pilot.class).iterator();		
		Pilot reloaded = (Pilot)next(storedObjects);

		Assert.isFalse(storedObjects.hasNext());
		
		reference = a().provider().produceReferenceByUUID(uuid, object1.getClass());
		Assert.areEqual(reference, a().provider().produceReference(reloaded, null, null));

		reloaded.setName("i am updated");
		a().provider().storeReplica(reloaded);

		a().provider().clearAllReferences();

		commitReplication();

		startReplication();

		reference = a().provider().produceReferenceByUUID(uuid, reloaded.getClass());
		Assert.areEqual("i am updated", ((Pilot) reference.object()).name());

		commitReplication();

		a().provider().deleteAllInstances(Pilot.class);
		a().provider().commit();
	}

	private DrsUUID uuid(Object obj) {
		return a().provider().produceReference(obj, null, null).uuid();
	}

}
